### PR TITLE
Add read lock before msg file reading.

### PR DESCRIPTION
**Pull request by karajan1001**
fix: #91

### CODE REVIEW

**Comment by karajan1001** [comment]
Excuse me  @pmrowla ,
how can I make DVC CI include this Patch before it is released?

**Comment by pmrowla** [comment]
> Excuse me @pmrowla , how can I make DVC CI include this Patch before it is released?

You have to just test it yourself locally. I generally just activate my DVC virtualenv and then do `pip install . --no-deps` in dvc-task

**Comment by karajan1001** [comment]
> > Excuse me @pmrowla , how can I make DVC CI include this Patch before it is released?
> 
> You have to just test it yourself locally. I generally just activate my DVC virtualenv and then do `pip install . --no-deps` in dvc-task

Yes, but it would be hard for me to test it with different version Python on various platforms.

**Comment by skshetry** [comment]
> Yes, but it would be hard for me to test it with different version Python on various platforms.

@karajan1001, you can change this:
https://github.com/iterative/dvc/blob/ed87c308ba0bb8780e59213343dcf049a6bcab91/setup.cfg#L68

to:
```
    dvc-task@git+https://github.com/iterative/dvc-task.git@fix91
```

File: src/dvc_task/app/filesystem.py [hunk 1]
```diff
@@ -122,6 +126,7 @@ def _iter_folder(
                     path = os.path.join(folder, filename)
                     try:
                         with open(path, "rb") as fobj:
+                            lock(fobj, LOCK_SH)
```
Comment (pmrowla) [inline src/dvc_task/app/filesystem.py:129]: does this need a matching `unlock()` call?

Comment (karajan1001) [reply src/dvc_task/app/filesystem.py:129]: > does this need a matching `unlock()` call?

> fobj.close() will automatically unlock it.

File: src/dvc_task/app/filesystem.py [hunk 2]
```diff
@@ -122,12 +127,16 @@ def _iter_folder(
...
                             payload = fobj.read()
+                            unlock(fobj)
```
Comment (pmrowla) [inline src/dvc_task/app/filesystem.py:132]:
```suggestion
                            lock(fobj, LOCK_SH)
                            try:
                                payload = fobj.read()
                            finally:
                                unlock(fobj)
```

**Review by pmrowla — APPROVED** [review]

**Comment by skshetry** [comment]
@karajan1001, can you please provide more information in the PR description above?

I don't really know what was broken, and how a lock on a read fixes this. Thanks. :)

**Comment by karajan1001** [comment]
> @karajan1001, can you please provide more information in the PR description above?
> 
> I don't really know what was broken, and how a lock on a read fixes this. Thanks. :)

Hi skshetry
The problem comes from the flaky CI test in the core repo. before this PR, we had a probability to read an incomplete message during the CI test.

**Comment by skshetry** [comment]
>  we had a probability to read an incomplete message during the CI test.

`write` should be atomic, no? How does a lock on `read()` fix that?

**Pull request merged**